protocol/x/prices/keeper/smoothed_price.go

```diff
@@ -1,33 +1,53 @@
 package keeper
 
 import (
+	"errors"
+	"fmt"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	"github.com/dydxprotocol/v4-chain/protocol/lib"
 	"github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
 )
 
 // UpdateSmoothedPrices updates the internal map of smoothed prices for all markets.
 // The smoothing is calculated with Basic Exponential Smoothing, see
 // https://en.wikipedia.org/wiki/Exponential_smoothing
 // If there is no valid index price for a market at this time, the smoothed price does not change.
-func (k Keeper) UpdateSmoothedPrices(ctx sdk.Context) error {
+func (k Keeper) UpdateSmoothedPrices(
+	ctx sdk.Context,
+	linearInterpolateFunc func(v0 uint64, v1 uint64, ppm uint32) (uint64, error),
+) error {
 	allMarketParams := k.GetAllMarketParams(ctx)
 	indexPrices := k.indexPriceCache.GetValidMedianPrices(allMarketParams, k.timeProvider.Now())
 
-	for market, indexPrice := range indexPrices {
-		smoothed, ok := k.marketToSmoothedPrices.GetSmoothedPrice(market)
+	// Track errors for each market.
+	updateErrors := make([]error, 0)
+
+	// Iterate through allMarketParams instead of indexPrices to ensure that we generate deterministic error messages
+	// in the case of failed updates.
+	for _, marketParam := range allMarketParams {
+		indexPrice, exists := indexPrices[marketParam.Id]
+		if !exists {
+			continue
+		}
+
+		smoothed, ok := k.marketToSmoothedPrices.GetSmoothedPrice(marketParam.Id)
 		if !ok {
 			smoothed = indexPrice
 		}
-		update, err := lib.Uint64LinearInterpolate(
+		update, err := linearInterpolateFunc(
 			smoothed,
 			indexPrice,
 			types.PriceSmoothingPpm,
 		)
 		if err != nil {
-			return err
+			updateErrors = append(
+				updateErrors,
+				fmt.Errorf("Error updating smoothed price for market %v: %w", marketParam.Id, err),
+			)
+			continue
 		}
-		k.marketToSmoothedPrices.PushSmoothedPrice(market, update)
+
+		k.marketToSmoothedPrices.PushSmoothedPrice(marketParam.Id, update)
 	}
-	return nil
+
+	return errors.Join(updateErrors...)
```

Comment:
to confirm, does this return nil if it's an empty list?

Reply:
It does! Existing test cases confirm this.